config/config-network.yaml

```diff
@@ -83,3 +83,13 @@ data:
     # update this value during the setup of Knative, to avoid getting
     # undefined behavior.
     clusteringress.class: "istio.ingress.networking.knative.dev"
+
+    # TLS mode of Knative. There are three TLS modes: "MANUAL", "AUTO", and "FORCE".
```

Comment:
I wonder if this would be better expressed through multiple knobs. There are many concepts being rolled into this setting, and not all combinations are possible with the three modes.

This setting seems to encapsulate:
1. Is Gateway reconciled
2. Are TLS certificates provisioned by Knative
3. Is HTTP or HTTP+HTTPs hosted
4. Does HTTP redirect to HTTPS

It seems like gateway reconciliation is a requirement of automatic TLS for us so perhaps that is combined into TLS enablement for now. The combination of endpoints and redirect settings though seem like they could be split out. Something like:

```
# Controls whether TLS certificates are provisioned. When enabled the ingress gateway is reconciled by the 
# Knative controller. Out of band changes will be overwritten. Default Disabled.
autoTLS: [Enabled|Disabled]
# Requires autoTLS to be enabled. Controls the behavior of the HTTP endpoint for the cluster. Default Enabled.
httpProtocol: [Enabled|Disabled|Redirected]
```

Reply:
I had a similar thought before.
But I would prefer using a single entry (`tlsMode`) to control TLS behavior as it is simple and straightforward.
I am afraid using the combination of `autoTLS` and `httpProtocol` could possibly confuse users.

Reply:
Grouping functionality into a single enum makes it difficult to expand later without breaking changes or an explosion of options. Lets chat about this offline as I think it will be faster.

Reply:
I am OK with Dan's proposal here considering its extensibility and flexibility.
Note that one scenario that was not covered by `tlsMode` is to support `HTTPS only` (i.e. `httpProtocol: Disabled` in Dan's proposal). I think it is reasonable to support `HTTPS only` feature as redirecting HTTP to HTTPS may not be always preferable for some users as 1) it takes extra request and 2) the first HTTP request may be hijacked.

WDYT @mattmoor @rgregg  ?

Reply:
I have a similar concern that autoTLS enabling gateway reconciliation is leaking implementation details through our abstraction layer. IMO it would be nicer if enable-gateway-reconciliation is an istio specific config. I think its reasonable if that override autoTLS when set to false and logged an error. This has the added benefit of us changing the default behavior separately when we feel like we can break istio < 1.1 by default (make gateway reconciliation default to on).

I wouldnt mind throwing up a follow on patch for this if this sounds reasonable

Reply:
@greghaynes having `enable-gateway-reconcilation` sounds good to me.

In this PR, I will mainly focus on auto-tls related configuration. Will follow up with `enable-gateway-reconcilation` in a separate PR.

Reply:
Now I updated the code to use `autoTLS` and `httpProtocol` to control auto-TLS. Please take a look.